File: atms/java/atms/Unify.java
// -*- Mode: Java -*-
//
// Variables and unification.
// Translated from unify.lisp, last edited 1/29/93 by KDF.
//
// Copyright (c) 1988-1993, Kenneth D. Forbus, Northwestern University,
// and Johan de Kleer, the Xerox Corporation.
// All rights reserved.
//
// See the file legal.txt for a paragraph stating scope of permission
// and disclaimer of warranty. The above copyright notice and that
// paragraph must be included in any separate copy of this file.

package atms;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Variables and unification.
 *
 * <p>Provides basic unification of s-expressions represented as
 * nested {@link Cons} cells and Java objects.  A <em>variable</em> is
 * any {@code String} whose first character is {@code '?'}.
 *
 * <p>Bindings are represented as a {@code List<Binding>} (an
 * association list).  A {@code null} return from {@link #unify}
 * with {@code success == false} signals failure (the Lisp
 * {@code :FAIL} sentinel).  See {@link UnifyResult} for details.
 */
public class Unify {

    // ---------------------------------------------------------------
    // Cons — a classic car/cdr pair for s-expression representation.
    // ---------------------------------------------------------------

    /**
     * A Lisp-style cons cell holding a car and a cdr.
     * Used to represent s-expressions (nested list structure).
     */
    public static class Cons {
        private final Object car;
        private final Object cdr;

        public Cons(Object car, Object cdr) {
            this.car = car;
            this.cdr = cdr;
        }

        public Object getCar() { return car; }
        public Object getCdr() { return cdr; }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Cons)) return false;
            Cons other = (Cons) o;
            return Objects.equals(car, other.car)
                && Objects.equals(cdr, other.cdr);
        }

        @Override
        public int hashCode() {
            return Objects.hash(car, cdr);
        }

        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder("(");
            sb.append(car);
            Object rest = cdr;
            while (rest instanceof Cons) {
                sb.append(' ');
                sb.append(((Cons) rest).car);
                rest = ((Cons) rest).cdr;
            }
            if (rest != null) {
                sb.append(" . ");
                sb.append(rest);
            }
            sb.append(')');
            return sb.toString();
        }
    }

    // ---------------------------------------------------------------
    // Binding — a single var → value pair.
    // ---------------------------------------------------------------

    /**
     * A binding of a variable name to a value.
     * Corresponds to a {@code (cons var exp)} entry in the Lisp
     * association-list of bindings.
     */
    public static class Binding {
        private final String var;
        private final Object val;

        public Binding(String var, Object val) {
            this.var = var;
            this.val = val;
        }

        public String getVar() { return var; }
        public Object getVal() { return val; }

        @Override
        public String toString() {
            return "(" + var + " . " + val + ")";
        }
    }

    // ---------------------------------------------------------------
    // UnifyResult — wrapper distinguishing success from failure.
    // ---------------------------------------------------------------

    /**
     * Result of a unification attempt.
     *
     * <p>In the original Lisp code, success with no bindings is
     * represented by the empty list {@code NIL}, while failure is
     * the keyword {@code :FAIL}.  Both would map to {@code null} in
     * Java, so this wrapper carries a boolean {@link #success} flag
     * to distinguish the two cases.
     */
    public static class UnifyResult {
        private final boolean success;
        private final List<Binding> bindings;

        private UnifyResult(boolean success, List<Binding> bindings) {
            this.success = success;
            this.bindings = bindings;
        }

        /** Successful unification with the given bindings. */
        public static UnifyResult success(List<Binding> bindings) {
            return new UnifyResult(true, bindings == null ? new ArrayList<>() : bindings);
        }

        /** Failed unification (corresponds to Lisp {@code :FAIL}). */
        public static UnifyResult fail() {
            return new UnifyResult(false, null);
        }

        public boolean isSuccess() { return success; }

        /**
         * Returns the list of bindings, or {@code null} if the
         * unification failed.
         */
        public List<Binding> getBindings() { return bindings; }

        @Override
        public String toString() {
            return success ? "Success" + (bindings == null ? "" : bindings.toString())
                           : ":FAIL";
        }
    }

    // ---------------------------------------------------------------
    // Public API
    // ---------------------------------------------------------------

    /**
     * Tests whether {@code x} is a variable — a {@code String}
     * whose first character is {@code '?'}.
     */
    public static boolean isVariable(Object x) {
        return (x instanceof String)
            && ((String) x).length() > 0
            && ((String) x).charAt(0) == '?';
    }

    /**
     * Unify two s-expressions {@code a} and {@code b} with no
     * initial bindings.
     */
    public static UnifyResult unify(Object a, Object b) {
        return unify(a, b, new ArrayList<>());
    }

    /**
     * Unify {@code a} and {@code b} under the given {@code bindings}.
     *
     * <p>Returns a {@link UnifyResult} indicating success (with
     * possibly extended bindings) or failure.
     *
     * <p>Corresponds to the Lisp function:
     * <pre>
     * (defun unify (a b &amp;optional (bindings nil)) ...)
     * </pre>
     */
    public static UnifyResult unify(Object a, Object b, List<Binding> bindings) {
        // (equal a b) => bindings
        if (Objects.equals(a, b)) {
            return UnifyResult.success(bindings);
        }
        // (variable? a) => unify-variable
        if (isVariable(a)) {
            return unifyVariable((String) a, b, bindings);
        }
        // (variable? b) => unify-variable
        if (isVariable(b)) {
            return unifyVariable((String) b, a, bindings);
        }
        // (or (not (listp a)) (not (listp b))) => :FAIL
        if (!(a instanceof Cons) || !(b instanceof Cons)) {
            return UnifyResult.fail();
        }
        // Recurse on car, then cdr
        Cons ca = (Cons) a;
        Cons cb = (Cons) b;
        UnifyResult carResult = unify(ca.car, cb.car, bindings);
        if (!carResult.isSuccess()) {
            return UnifyResult.fail();
        }
        return unify(ca.cdr, cb.cdr, carResult.getBindings());
    }

    /**
     * Attempt to bind {@code var} to {@code exp} under
     * {@code bindings}.
     *
     * <p>Corresponds to the Lisp function:
     * <pre>
     * (defun unify-variable (var exp bindings) ...)
     * </pre>
     */
    public static UnifyResult unifyVariable(String var, Object exp, List<Binding> bindings) {
        Binding val = assoc(var, bindings);
        if (val != null) {
            // Variable already bound — unify bound value with exp.
            return unify(val.val, exp, bindings);
        }
        // If safe (no occurs-check violation), bind var to exp.
        if (freeIn(var, exp, bindings)) {
            List<Binding> extended = new ArrayList<>(bindings);
            extended.add(new Binding(var, exp));
            return UnifyResult.success(extended);
        }
        return UnifyResult.fail();
    }

    /**
     * Returns {@code true} if {@code var} does <em>not</em> occur in
     * {@code exp} (the "occurs check").
     *
     * <p>Corresponds to the Lisp function:
     * <pre>
     * (defun free-in? (var exp bindings) ...)
     * </pre>
     */
    public static boolean freeIn(Object var, Object exp, List<Binding> bindings) {
        if (exp == null) {
            return true;
        }
        if (Objects.equals(var, exp)) {
            return false;
        }
        if (isVariable(exp)) {
            Binding val = assoc((String) exp, bindings);
            if (val != null) {
                return freeIn(var, val.val, bindings);
            }
            return true;
        }
        if (!(exp instanceof Cons)) {
            return true;
        }
        Cons c = (Cons) exp;
        return freeIn(var, c.car, bindings)
            && freeIn(var, c.cdr, bindings);
    }

    // ---------------------------------------------------------------
    // Helpers
    // ---------------------------------------------------------------

    /**
     * Look up {@code var} in the binding list (association list).
     * Returns the first {@link Binding} whose var matches, or
     * {@code null} if not found.
     */
    public static Binding assoc(String var, List<Binding> bindings) {
        if (bindings == null) return null;
        for (Binding b : bindings) {
            if (b.var.equals(var)) {
                return b;
            }
        }
        return null;
    }

    // ---------------------------------------------------------------
    // Convenience builders for Cons lists
    // ---------------------------------------------------------------

    /** Build a proper list from the given elements. */
    public static Cons list(Object... elements) {
        Cons result = null;
        for (int i = elements.length - 1; i >= 0; i--) {
            result = new Cons(elements[i], result);
        }
        return result;
    }

    /** Build a cons cell. */
    public static Cons cons(Object car, Object cdr) {
        return new Cons(car, cdr);
    }

    /** Return the car of a Cons, or null. */
    public static Object car(Object x) {
        return (x instanceof Cons) ? ((Cons) x).car : null;
    }

    /** Return the cdr of a Cons, or null. */
    public static Object cdr(Object x) {
        return (x instanceof Cons) ? ((Cons) x).cdr : null;
    }
}

File: atms/go/unify.go
// -*- Mode: Go; -*-

// Variables and unification
// Last edited 1/29/93, by KDF

// Copyright (c) 1988-1993, Kenneth D. Forbus, Northwestern University,
// and Johan de Kleer, the Xerox Corporation.
// All rights reserved.

// See the file legal.txt for a paragraph stating scope of permission
// and disclaimer of warranty. The above copyright notice and that
// paragraph must be included in any separate copy of this file.

package atms

import "reflect"

// Binding represents an association of a variable to a value,
// corresponding to a Lisp (var . value) cons cell in the bindings alist.
type Binding struct {
	Var string
	Val interface{}
}

// Cons represents a Lisp cons cell for building s-expressions.
type Cons struct {
	Car interface{}
	Cdr interface{}
}

// IsVariable returns true if x is a string starting with "?".
func IsVariable(x interface{}) bool {
	s, ok := x.(string)
	return ok && len(s) > 0 && s[0] == '?'
}

// Unify attempts to unify two s-expressions a and b given existing bindings.
// Returns the extended bindings and true on success, or nil and false on failure.
func Unify(a, b interface{}, bindings []Binding) ([]Binding, bool) {
	if SExprEqual(a, b) {
		return bindings, true
	}
	if IsVariable(a) {
		return unifyVariable(a.(string), b, bindings)
	}
	if IsVariable(b) {
		return unifyVariable(b.(string), a, bindings)
	}
	aCons, aOk := a.(*Cons)
	bCons, bOk := b.(*Cons)
	if !aOk || !bOk {
		return nil, false
	}
	bindings, ok := Unify(aCons.Car, bCons.Car, bindings)
	if !ok {
		return nil, false
	}
	return Unify(aCons.Cdr, bCons.Cdr, bindings)
}

// unifyVariable attempts to unify a variable with an expression.
// Must distinguish no value from a value of nil.
func unifyVariable(v string, exp interface{}, bindings []Binding) ([]Binding, bool) {
	val, found := assoc(v, bindings)
	if found {
		return Unify(val, exp, bindings)
	}
	// If safe, bind v to exp.
	if FreeIn(v, exp, bindings) {
		return append(bindings, Binding{Var: v, Val: exp}), true
	}
	return nil, false
}

// FreeIn returns true if var does not occur in exp, assuming the given bindings.
func FreeIn(v string, exp interface{}, bindings []Binding) bool {
	if exp == nil {
		return true
	}
	if SExprEqual(v, exp) {
		return false
	}
	if IsVariable(exp) {
		val, found := assoc(exp.(string), bindings)
		if found {
			return FreeIn(v, val, bindings)
		}
		return true
	}
	cons, ok := exp.(*Cons)
	if !ok {
		return true
	}
	if FreeIn(v, cons.Car, bindings) {
		return FreeIn(v, cons.Cdr, bindings)
	}
	return false
}

// assoc looks up a variable in the bindings list, returning its value and
// whether it was found.
func assoc(v string, bindings []Binding) (interface{}, bool) {
	for _, b := range bindings {
		if b.Var == v {
			return b.Val, true
		}
	}
	return nil, false
}

// SExprEqual performs deep equality comparison on s-expressions,
// handling Cons cells, slices, and primitive types.
func SExprEqual(a, b interface{}) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	aCons, aOk := a.(*Cons)
	bCons, bOk := b.(*Cons)
	if aOk && bOk {
		return SExprEqual(aCons.Car, bCons.Car) && SExprEqual(aCons.Cdr, bCons.Cdr)
	}
	if aOk != bOk {
		return false
	}
	return reflect.DeepEqual(a, b)
}
